### PR TITLE
[CSL-2176] Respect some invariants

### PR DESCRIPTION
**Pull request by neongreen**
I recommend reviewing each commit individually.

### CODE REVIEW

File: crypto/Pos/Crypto/Signing/Types/Signing.hs [hunk 1]
```diff
@@ -140,7 +140,9 @@ instance B.Buildable (ProxyCert w) where
 
 -- | Convenient wrapper for secret key, that's basically ω plus
 -- certificate.
-data ProxySecretKey w = ProxySecretKey
+--
+-- The invariant is that the certificate is valid.
```
Comment (volhovm) [inline crypto/Pos/Crypto/Signing/Types/Signing.hs]: What is "valid"? You should add "with respect to `validateProxySecretKey`" I think.

File: core/Pos/Core/Genesis/Canonical.hs [hunk 2]
```diff
@@ -386,7 +386,7 @@ instance ReportSchemaErrors m => FromJSON m (ProxySecretKey EpochIndex) where
         pskIssuerPk <- fromJSField obj "issuerPk"
         pskDelegatePk <- fromJSField obj "delegatePk"
         pskCert <- fromJSField obj "cert"
-        return ProxySecretKey {..}
+        pure UnsafeProxySecretKey{..}
```
Comment (volhovm) [inline core/Pos/Core/Genesis/Canonical.hs]: Why are you sure that this psk is valid?

File: delegation/src/Pos/Delegation/Logic/Mempool.hs [hunk 3]
```diff
@@ -195,7 +200,6 @@ processProxySKHeavyInternal psk = do
 
         let res = if | cached -> PHCached
                      | not coherent -> PHTipMismatch
-                     | not consistent -> PHBroken
```
Comment (volhovm) [inline delegation/src/Pos/Delegation/Logic/Mempool.hs]: You should delete the related constructor also.

File: crypto/Pos/Crypto/Signing/Signing.hs [hunk 4]
```diff
@@ -8,21 +8,20 @@ module Pos.Crypto.Signing.Signing
...
+       , checkSig                      -- reexport
        , fullSignatureHexF
        , parseFullSignature
-
        , mkSigned
 
        -- * Versions for raw bytestrings
        , signRaw
-       , verifyRaw
+       , checkSigRaw                   -- reexport
```
Comment (volhovm) [inline crypto/Pos/Crypto/Signing/Signing.hs]: So all of this was duplicated? ..
Does https://iohk.myjetbrains.com/youtrack/issue/CSL-1157 have any relation to it?

File: update/Pos/Update/Logic/Local.hs [hunk 5]
```diff
@@ -395,6 +395,8 @@ finishPrepare badProposals proposals votes = do
...
         | otherwise =
             deactivateProposal upId
-    isVoteValid UpdateVote {..} =
-        (not (HS.member uvProposalId badProposals) &&) . isJust <$>
-        getProposal uvProposalId
+    isVoteValid vote = do
+        let id = uvProposalId vote
+        proposalIsPresent <- isJust <$> getProposal id
+        pure $ not (HS.member id badProposals) && proposalIsPresent
+
```
Comment (volhovm) [inline update/Pos/Update/Logic/Local.hs]: Trailing line.

File: crypto/Pos/Crypto/Signing/Signing.hs (hunk 4 above)
Comment (neongreen) [reply crypto/Pos/Crypto/Signing/Signing.hs]: No relation.

I'm not sure what you mean by “duplicated”.

File: core/Pos/Core/Genesis/Canonical.hs (hunk 2 above)
Comment (neongreen) [reply core/Pos/Core/Genesis/Canonical.hs]: I'm not sure it's valid, but stuff like `VssCertificate` isn't checked in `Pos.Core.Genesis.Canonical` either so I assumed that whoever wrote `Pos.Core.Genesis.Canonical` knew what they were doing. Maybe it was a wrong assumption in hindsight. If you want, I can change it.

**Comment by volhovm** [comment]
I need these changes to implement CSL-2072 inside PR #2324, so I'm merging it.

**Pull request merged**